*** src/Copie de jsutil.h	Tue Sep 22 17:17:30 2009
--- src/jsutil.h	Fri Sep 25 13:27:42 2009
***************
*** 49,52 ****
--- 49,70 ----
  JS_BEGIN_EXTERN_C
  
+ typedef void* (*jslibs_malloc_t)( size_t size );
+ typedef void* (*jslibs_calloc_t)( size_t num, size_t size );
+ typedef void* (*jslibs_realloc_t)( void *ptr, size_t size );
+ typedef void (*jslibs_free_t)( void *ptr );
+ 
+ extern jslibs_malloc_t jslibs_malloc;
+ extern jslibs_calloc_t jslibs_calloc;
+ extern jslibs_realloc_t jslibs_realloc;
+ extern jslibs_free_t jslibs_free;
+ 
+ extern JS_PUBLIC_API(void)
+ JSLIBS_RegisterAllocFunctions( jslibs_malloc_t malloc, jslibs_calloc_t calloc, jslibs_realloc_t realloc, jslibs_free_t free );
+ 
+ #define malloc jslibs_malloc
+ #define calloc jslibs_calloc
+ #define realloc jslibs_realloc
+ #define free jslibs_free
+ 
  /*
   * JS_Assert is present even in release builds, for the benefit of applications
***************
*** 204,207 ****
--- 222,230 ----
  
  JS_END_EXTERN_C
+ 
+ #undef malloc
+ #undef calloc
+ #undef realloc
+ #undef free
  
  #endif /* jsutil_h___ */
