--- 
+++ src/js/public/Utility.h	2013-06-24 17:53:34 +0200
@@ -79,8 +79,25 @@
 /*
  * Custom allocator support for SpiderMonkey
  */
-#if defined JS_USE_CUSTOM_ALLOCATOR
-# include "jscustomallocator.h"
+#if 1
+#define HAS_JL_ALLOCATORS
+extern JS_PUBLIC_DATA(uint32_t) OOM_maxAllocations;
+extern JS_PUBLIC_DATA(uint32_t) OOM_counter;
+#define JS_OOM_POSSIBLY_FAIL() do {} while(0)
+#define JS_OOM_POSSIBLY_FAIL_REPORT(cx) do {} while(0)
+typedef void* (*js_jl_malloc_t)( size_t );
+typedef void* (*js_jl_calloc_t)( size_t, size_t );
+typedef void* (*js_jl_realloc_t)( void*, size_t );
+typedef void (*js_jl_free_t)( void* );
+extern JS_PUBLIC_DATA(js_jl_malloc_t) js_jl_malloc;
+extern JS_PUBLIC_DATA(js_jl_calloc_t) js_jl_calloc;
+extern JS_PUBLIC_DATA(js_jl_realloc_t) js_jl_realloc;
+extern JS_PUBLIC_DATA(js_jl_free_t) js_jl_free;
+static inline void* js_malloc(size_t size) { return js_jl_malloc(size); }
+static inline void* js_calloc(size_t size, size_t count = 1) { return js_jl_calloc(size, count); }
+static inline void* js_realloc(void* ptr, size_t size) { return js_jl_realloc(ptr, size); }
+static inline void js_free(void* ptr) { js_jl_free(ptr); }
+// /HAS_JL_ALLOCATORS
 #else
 # ifdef DEBUG
 /*
